homeassistant/components/homekit/type_sensors.py

```diff
@@ -128,21 +146,194 @@ def __init__(self, *args):
+            or DEVICE_CLASS_PM25 in self.entity_id
+        ):
+            chars.append(CHAR_AIR_PARTICULATE_DENSITY)
+
+        if (
+            pm_size is not None
+            or device_class == DEVICE_CLASS_PM25
+            or DEVICE_CLASS_PM25 in self.entity_id
+        ):
+            chars.append(CHAR_AIR_PARTICULATE_SIZE)
+
+        _LOGGER.debug("%s: chars: %s", self.entity_id, chars)
+
+        # add the service with the determined characteristics
+        serv_air_quality = self.add_preload_service(SERV_AIR_QUALITY_SENSOR, chars)
+        _LOGGER.debug("%s, service: %s", self.entity_id, serv_air_quality.to_HAP())
+
+        # configure the determined characteristics
+        self.char_air_quality = serv_air_quality.configure_char(
+            CHAR_AIR_QUALITY, value=0
         )
 
+        if CHAR_NITROGEN_DIOXIDE_DENSITY in chars:
+            self.char_nitrogen_dioxide_density = serv_air_quality.configure_char(
+                CHAR_NITROGEN_DIOXIDE_DENSITY, value=0
+            )
+
+        if CHAR_VOC_DENSITY in chars:
+            self.char_voc_density = serv_air_quality.configure_char(
+                CHAR_VOC_DENSITY, value=0
+            )
+
+        if CHAR_PM_2_5_DENSITY in chars:
+            self.char_pm_2_5_density = serv_air_quality.configure_char(
+                CHAR_PM_2_5_DENSITY, value=0
+            )
+
+        if CHAR_PM_10_DENSITY in chars:
+            self.char_pm_10_density = serv_air_quality.configure_char(
+                CHAR_PM_10_DENSITY, value=0
+            )
+
+        if CHAR_AIR_PARTICULATE_DENSITY in chars:
+            self.char_particulate_density = serv_air_quality.configure_char(
+                CHAR_AIR_PARTICULATE_DENSITY, value=0
+            )
+
+        if CHAR_AIR_PARTICULATE_SIZE in chars:
+            self.char_particulate_size = serv_air_quality.configure_char(
+                CHAR_AIR_PARTICULATE_SIZE, value=CHAR_VALUE_AIR_PARTICULATE_SIZE_PM2_5
+            )
+
     def update_state(self, new_state):
         """Update accessory after state change."""
-        density = convert_to_float(new_state.state)
-        if density:
-            self.char_density.set_value(density)
-            self.char_quality.set_value(density_to_air_quality(density))
-            _LOGGER.debug("%s: Set to %d", self.entity_id, density)
+        nitrogen_dioxide_density = convert_to_float(
+            new_state.attributes.get(ATTR_NITROGEN_DIOXIDE_DENSITY)
+        )
+        voc_density = convert_to_float(new_state.attributes.get(ATTR_VOC_DENSITY))
+        pm_2_5_density = convert_to_float(new_state.attributes.get(ATTR_PM_2_5_DENSITY))
+        pm_10_density = convert_to_float(new_state.attributes.get(ATTR_PM_10_DENSITY))
+        pm_density = convert_to_float(new_state.attributes.get(ATTR_PM_DENSITY))
+        pm_size = convert_to_float(new_state.attributes.get(ATTR_PM_SIZE))
+
+        # if device class or entity ID contain DEVICE_CLASS_PM25 then assume state is a PM25 reading
```

Comment:
Looking at the [Air Quality Entity docs](https://developers.home-assistant.io/docs/en/entity_air_quality.html) and the [air quality entity code](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/air_quality/__init__.py) leads me to not be sure about about this.

It looks like `state` for a correctly implemented Air Quality Entity would be the pm2_5, and the homekit code should really be aiming to support that in the first instance.

It looks like there is an `ATTR_AQI` field in the entity state you could use to derive the air quality, or fall back to the existing `density_to_air_quality`, but you probably shouldn't rely on HomeKit specific values even if you can create them yourself with a template sensor.

Reply:
> Looking at the Air Quality Entity docs and the air quality entity code leads me to not be sure about about this.

I knew that my dyson fan created an `air_quality.` domain entity, but I didn't understand it was an actually defined thing in Home Assistant. I thought it was some just custom entity to dump all the data in that the Dyson integration was creating. Now that I understand that `air_quality` is a defined entity domain...

That brings up an entirely new thought. Rather then (or in addition to) looking for `air_quality` in the device type or the entit_id of a `sensor` domain. Should instead (or in addition to) should we b looking for any `air_quality` domain entities and creating an `AirQuaitySensor` from that?

> It looks like state for a correctly implemented Air Quality Entity would be the pm2_5, and the homekit code should really be aiming to support that in the first instance.

That seems really weird. Shouldn't the state of an Air Quality Entity be using the Air Quality Index as its state and not one (seemingly arbitrary) property from the entity?

> It looks like there is an ATTR_AQI field in the entity state you could use to derive the air quality, or fall back to the existing density_to_air_quality, but you probably shouldn't rely on HomeKit specific values even if you can create them yourself with a template sensor.

You lost me here. Are you refering to the old code path that did the pm2.5 to air quality conversion or something I am doing in the new path?

With my custom sensor templates the reason I have those blocks to convert each individual air quality property to a home kit air quality value [`HMCharacteristicValueAirQuality`](https://developer.apple.com/documentation/homekit/hmcharacteristicvalueairquality) is so I can have different sensors in HomeKit that report the air quality based on the different properties of the air quality as well as an aggregate sensor that reports the air quality based on aggregate of all the properties. Did this because the summary of the air quality sensor in HomeKit only shows the [`HMCharacteristicValueAirQuality`](https://developer.apple.com/documentation/homekit/hmcharacteristicvalueairquality) and you have to click in to see any property values and wanted an "at a glance" look at the level of each property.

Reply:
> That brings up an entirely new thought. Rather then (or in addition to) looking for air_quality in the device type or the entit_id of a sensor domain. Should instead (or in addition to) should we b looking for any air_quality domain entities and creating an AirQuaitySensor from that?

I'm not the maintainer for this integration so i'm going to have to make a tooth sucking noise and say maybe? I would certainly expect it to create a homekit proxy for any entity in the air_quality domain I guess?

> That seems really weird. Shouldn't the state of an Air Quality Entity be using the Air Quality Index as its state and not one (seemingly arbitrary) property from the entity?

I'm just going off what the code in `homeassistant/components/air_quality/__init__.py` currently does. Changing that would need an architecture issue and would technically be a breaking change.

> You lost me here. Are you refering to the old code path that did the pm2.5 to air quality conversion or something I am doing in the new path?

On line 250 and line 269 you use new_state.state. You use it as a pm2.5 in the old path and as a air quality in the other. What i meant was the type of data in .state isn't very well documented but seems to always be the pm2.5 value for a `air_quality` domain entity. So even though it's convenient for your use case it's probably "wrong" for the average HA user. I like the idea of doing the mapping from the state.attributes[ATTR_AQI] field to a HMCharacteristicValueAirQuality though.